internal/messaging/response_handler.go

```diff
@@ -1,147 +1,536 @@
+	switch msgType { //nolint:gocritic
+	case MintRequest:
+		request.BuyerAddress = crypto.PubkeyToAddress(h.pk.ToECDSA().PublicKey).Hex()
+	}
+	return nil
+}
+
+func (h *EvmResponseHandler) handleMintResponse(_ context.Context, response *ResponseContent, request *RequestContent) bool {
 	if response.MintResponse.Header == nil {
 		response.MintResponse.Header = &typesv1alpha.ResponseHeader{}
 	}
-	buyer, err := codec.ParseAddressBech32(consts.HRP, request.MintRequest.BuyerAddress)
+	bookingTokenABIFile := h.cfg.BookingTokenABIFile
+	abi, err := loadABI(bookingTokenABIFile)
+	if err != nil {
+		addErrorToResponseHeader(response, fmt.Sprintf("error loading ABI: %v", err))
+		return true
+	}
+	address := crypto.PubkeyToAddress(h.pk.ToECDSA().PublicKey)
+
+	bookingTokenAddress := common.HexToAddress(h.cfg.BookingTokenAddress)
+
+	packedData, err := abi.Pack("getSupplierName", address)
+	if err != nil {
+		h.logger.Infof("Error packing data for getSupplierName: %v", err)
+		addErrorToResponseHeader(response, fmt.Sprintf("error packing data: %v", err))
+		return true
+	}
+
+	msg := ethereum.CallMsg{
+		To:   &bookingTokenAddress,
+		Data: packedData,
+	}
+	result, err := h.ethClient.CallContract(context.Background(), msg, nil)
 	if err != nil {
-		addErrorToResponseHeader(response, fmt.Sprintf("error parsing buyer address: %v", err))
+		h.logger.Infof("Error calling contract: %v", err)
+		addErrorToResponseHeader(response, fmt.Sprintf("error calling contract: %v", err))
 		return true
 	}
-	price, err := strconv.Atoi(response.MintResponse.Price.Value)
+
+	// Unpack the result
+	var supplierName string
+	err = abi.UnpackIntoInterface(&supplierName, "getSupplierName", result)
 	if err != nil {
-		addErrorToResponseHeader(response, fmt.Sprintf("error parsing price value: %v", err))
+		h.logger.Infof("Error unpacking result: %v", err)
+		addErrorToResponseHeader(response, fmt.Sprintf("error unpacking result: %v", err))
 		return true
 	}
-	success, txID, err := h.tvmClient.SendTxAndWait(ctx, createNFTAction(owner, buyer, uint64(response.MintResponse.BuyableUntil.Seconds), uint64(price), response.MintResponse.MintId))
+
+	if supplierName != h.cfg.SupplierName {
+		h.logger.Debugf("Not registered with correct supplier name: %v != %v", supplierName, h.cfg.SupplierName)
+		h.logger.Debugf("Registerring with supplierName: %v", h.cfg.SupplierName)
+		err := register(h.ethClient, abi, h.pk.ToECDSA(), h.cfg.SupplierName, bookingTokenAddress)
+		if err != nil {
+			addErrorToResponseHeader(response, fmt.Sprintf("error registering supplier: %v", err))
+			return true
+		}
+	} else {
+		h.logger.Debugf("Supplier is already registered with supplierName: %v", supplierName)
+	}
+	// TODO @evlekht ensure that request.MintRequest.BuyerAddress is c-chain address format, not x/p/t chain
+	buyer := common.HexToAddress(request.MintRequest.BuyerAddress)
+
+	h.logger.Debugf("abi: %v", abi)
+
+	// TODO @evlekht unhardocoded, figure out what it is at all
```

Comment:
I don't recall what it is, plz remove it >_<

Reply:
It was noted for the hardcoded token URI. I've created and used the `createTokenURIforMintResponse` function to generate token URIs (base64 encoded json, on-chain) based on the `MintResponse` data. 

We still have hardcoded info inside that, like image URL, name, description etc., but those are not relevant for this PR because we also need to update the protocol to have those information in the mint response.